### ISSUE TITLE
Missmatched Types and Methods

### ISSUE THREAD

**Issue by motleydev** [teslamotors/informed#392]
**Describe the bug**
I've been using Informed for a little over a week so this issue spans previous to the 4.2 release and is persistent after the 4.2 release. I've deleted my modules folder, and the problem persists.

Following the getting started guide, when I install Informed and attempt to import the different components, the majority complain about having no exported member matching the name.

![CleanShot 2022-02-12 at 12 38 15](https://user-images.githubusercontent.com/1203804/153709731-5ab9c60f-99b5-4a2e-b3af-9433c4487b09.png)

Using the formApi hook, and attempting to call `setValue` checks out from the types, but the method doesn't exist when actually attempted to be invoked.

It seems like there is some kind of large type mismatch/cache issue but I, unfortunately, don't know where else to look at this point.

**To Reproduce**

Scaffold a boilerplate NextJs project with typescript.
Following the getting started guide from Informed.
Try to use a formApi setValue method (or the documented setValues that typescript complains doesn't exist).

My files use the .tsx extension if that makes a difference?

**Comment by joepuzzo** [comment]
Can you try to recreate in a sandbox or something. I think you are the only person who has this issue.

**Comment by motleydev** [comment]
Sure! Following the instructions, I listed above. This is a brand new setup (Feb 15, 12pm CET)

TLDR; Repo: https://github.com/motleydev/sandbox-informed

1. Scaffold the app: `yarn create next-app --typescript`
2. Enter the directory, add informed `yarn add informed`
3. Open an index page from NextJs, paste in the imports: `import { Form, Input, Select, Checkbox, Relevant, Debug } from 'informed';

From there it already shows that the exports are missing (along with unused) - add the boilerplate code:

```js
`<Form onSubmit={onSubmit}>
    <Input name="name" label="Name" placeholder="Elon" />
    <Input name="age" type="number" label="Age" required="Age Required" />
    <Input name="phone" label="Phone" formatter="+1 (###)-###-####" />
    <Select name="car" label="Car" initialValue="ms">
      <option value="ms">Model S</option>
      <option value="m3">Model 3</option>
      <option value="mx">Model X</option>
      <option value="my">Model Y</option>
    </Select>
    <Checkbox name="married" label="Married?" />
    <Relevant when={({ formState }) => formState.values.married}>
      <Input name="spouse" label="Spouse" />
    </Relevant>
    <button type="submit">Submit</button>
    <Debug />
  </Form>
  ```
  And now it only complains about the missing exports.

<img width="901" alt="CleanShot 2022-02-15 at 11 57 20@2x" src="https://user-images.githubusercontent.com/1203804/154048481-2ba5f8f5-a713-4d78-bc52-8fcd82bb63fa.png">

  
Additionally, import the `useFormApi` hook and call it on the page:

`const formAPI = useFormApi()`

Add it to a button below the form.

```js
<button
          onClick={() => {
            formAPI.setValue("email", "foo@bar.baz");
          }}
        >
          Set Email to foo@bar.baz
        </button>
```

The types seem to be present, but calling the actual method yields an error.

<img width="1112" alt="CleanShot 2022-02-15 at 12 05 17@2x" src="https://user-images.githubusercontent.com/1203804/154050402-e811f20b-9848-444e-aac9-5790e9ce0be1.png">

**Comment by ghoshabhi** [comment]
@joepuzzo Looks like we're still missing types for the form field components (in 4.5.1)

@motleydev the error for `formApi.setValue` occurs because you are calling `useFormApi` from outside the form context. `informed` uses React context as the means to share data across components.

To achieve what you're trying to you can create a component that gets rendered as a child of `Form`:
```js
const SetValueButton = () => {
  const formAPI = useFormApi();
  return (
    <button
      type="button"
      onClick={() => {
        formAPI.setValue("email", "foo@bar.baz");
      }}
    >
      Set Email to foo@bar.baz
    </button>
  )
}

// render SetValueButton as a child of <Form>
<Form>
   ...
   <SetValueButton />
</Form>
```

Following works:
<img width="1356" alt="image" src="https://user-images.githubusercontent.com/19292575/154733194-b67e09f5-766f-4464-9c9f-fb49572d0706.png">

---
A side note: it looks like you were passing incorrect props for `required`, `required` is a boolean. It looks like docs are missing some of the props.

<img width="901" alt="image" src="https://user-images.githubusercontent.com/19292575/154733725-1d50e842-6b6d-4ef2-8477-c45900090430.png">

**Comment by motleydev** [comment]
@ghoshabhi Ah thanks! I guess that makes sense in context with a Form, its just not something I was familiar with. Is this explicit in the docs? That might help. If you "stop to think about it" then it makes sense, because, forms. But its a bit a-typical for a lot React lib APIs that don't have something like a "FormProvider" component. 

Re the required props, that's literally copy and paste from the website, so maybe another place to update. :) I'll try to poke around and see where that needs to get updated.

![CleanShot 2022-02-19 at 08 05 05@2x](https://user-images.githubusercontent.com/1203804/154790874-e7b6bd84-6fd1-40a1-81dc-90f68d2b6c10.png)

**Comment by joepuzzo** [comment]
Working on it now

**Comment by joepuzzo** [comment]
@motleydev This should all be fixed now in latest version.

**Comment by joepuzzo** [comment]
Actually I may have forgot to do Debug 🤦‍♂️

**Comment by motleydev** [comment]
@ghoshabhi it seems the latest update might have introduced a type regression on useFormApi, this is the list of available methods of `useFormApi` - if I recall your proposed solution as working in the previous version.

(Length 26)
getValue
setValue
getMaskedValue
setMaskedValue
getTouched
setTouched
getError
setError
getFocused
setFocused
getData
resetField
reset
getFormState
getPristine
getDirty
validateField
getFieldState
getInitialValue
touchAllFields
validate
asyncValidate
setValues
setTheseValues
submitForm
clearValue

These are the typed methods:

(Length 19)
getValue
setValue
getMaskedValue
setMaskedValue
getTouched
setTouched
getError
setError
getFocused
setFocused
resetField
reset
getFormState
getPristine
getDirty
getFieldState
validate
validateField
submitForm

**Comment by joepuzzo** [comment]
Yup looks like I missed a few types I will add and keep this open until I do add

**Comment by motleydev** [comment]
Hey, just checking on the current status of this? Have these been updated?

**Comment by joepuzzo** [comment]
Types have been updated many times since then please try with the latest version.
